### ISSUE TITLE
str vs. unicode

### ISSUE THREAD

**Issue by torfsen** [ckan/ckan#3006]
In Python 2, `__repr__` and `__str__` must return instances of `str`. Returning `unicode` instances fails:
```
Python 2.7.6 (default, Jun 22 2015, 18:00:18) 
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> class Foo(object):
...     def __str__(self): return u'ö'
...     def __repr__(self): return u'ö'
... 
>>> str(Foo())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'ascii' codec can't encode character u'\xf6' in position 0: ordinal not in range(128)
>>> repr(Foo())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'ascii' codec can't encode character u'\xf6' in position 0: ordinal not in range(128)
```
I ran into this problem on my CKAN 2.5.2 installation due to tags containing non-ASCII characters, because `model.Tag.__repr__` [looks like this](https://github.com/ckan/ckan/blob/master/ckan/model/tag.py#L215):
```
def __repr__(self):
    return '<Tag %s>' % self.name
```
That expression will return either a `str` or a `unicode` instance, depending on the type of `self.name`.

I tried to fix that (and similar) errors by making sure that all implementations of `__repr__` and `__str__` always return `str` and `__unicode__` always returns `unicode`. However, while they do enforce the correct behavior, these fixes break a lot of other pieces of code which (implicitly) assumed the old behavior.

The problem is that there seems to be no rule whether `str` or `unicode` are to be used in a given context. Therefore, each part of the code which does string processing and object-to-string conversion needs to make sure it handles both `str` and `unicode` correctly. The errors that pop up after I fixed `__repr__`, etc. show that this is often not done correctly.

Before spending a lot of time fixing each of these little issues I want to make sure that I'm on the right track concerning how Unicode is handled in CKAN. Correct handling of Unicode in Python 2 is a non-trivial issue, and in my opinion CKAN would benefit greatly from a clear strategy. This is especially true with regard to an eventual migration to Python 3. In particular:

- What encoding should CKAN Python source files use (see [PEP 263](https://www.python.org/dev/peps/pep-0263/))?
- Do we want to use `unicode_literals` from [`__future__`](https://docs.python.org/2/library/__future__.html)?
- Do we use `str` or `unicode` internally?
- In what form do we pass/receive strings to/from Pylons, babel, the file system, etc.?

Are there existing CKAN conventions with regard to this? The documentation only mentions Unicode in a few isolated places.

**Comment by wardi** [comment]
I'm really happy to see work done in this area. There are all sorts of simple failures like this that I shouldn't be seeing in my logs.

1. For source files our encoding should always be UTF-8.
2. I am slightly against unicode_literals because it means you need to scan to the top of the module to see what `'foo'` actually means, and moving a function from one module to another might break. However, for source compatibility with Python 3 it is a nice thing to have, so I could be convinced.
3. We never need to store binary strings in CKAN, all our variables should be unicode
4. jinja2: unicode, babel unicode (though I've seen binary strings coming returned, need to investigate), file system: str (file names are str utf-8 too since we're only supporting linux)

**Comment by amercader** [comment]
@torfsen You are absolutely right in that there is no consistency internally, mainly for historical reasons. I agree on all @wardi's points, mainly that source files should be encoded in UTF-8 and we should use unicode everywhere internally.

As for whether to use `unicode_literals` or not I must confess I didn't have much experience with it, but after reading this http://python-future.org/unicode_literals.html I'm more leaning towards not using it. What's the alternative though? Is it going through the code and marking all strings with `u'...'`? Is that feasible?

**Comment by torfsen** [comment]
Regarding `unicode_literals`:

The safe way is to not use `unicode_literals` and instead use a `u` prefix for all string literals. Python 3 versions before 3.3 don't have that prefix, but as far as I know 3.3 is usually considered the first Python 3 version that is reasonable to support nowadays. While we do have quite a lot of string literals in our code adding the necessary prefixes once is manageable with a suitable editor. The main problem, in my opinion, is that you have to keep adding the prefix for new literals, too. One would probably need some kind of automatic check to ensure that. In addition, any literal that deliberately hasn't got a `u` prefix would need a comment, since otherwise you'd never know whether the prefix was simply forgotten (with `unicode_literals`, you'd use the `b` prefix to get a `str` literal, something that wouldn't happen on accident).

The biggest problem with using `unicode_literals`, on the other hand, seems to be that you have to be careful when calling libraries and IO-functions because they might behave unexpectedly (or simply differently) when being passed Unicode data. In addition, one needs to make sure that `unicode_literals` is actually included in every module.

Basically it's a mess, no matter which way. Judging from my (limited) experience, using `unicode_literals` is easier because it does the right thing most of the time and because it's less obtrusive in day-to-day work.

**Comment by torfsen** [comment]
I'm going to collect some additional information regarding this issue and dump it here until we have a clearer strategy of how to proceed.

#### Jinja2
* According to the [official documentation](http://jinja.pocoo.org/docs/dev/api/#unicode), Jinja2 internally uses Unicode. In particular, rendering a template [returns Unicode](http://jinja.pocoo.org/docs/dev/api/#jinja2.Template.render).
* The docs also recommend to UTF8-encode Python modules and template files (the latter is assumed by default)
* To stay compatible with libraries that explicitly type check agains `str`, Jinja2 handles string literals *in templates* specially: If their content is ASCII-only then they are represented by `str` instances, otherwise `unicode` is used. There is currently no way to change this behavior (the `u` prefix is not supported in templates), and this can lead to [subtle issues](https://github.com/pallets/jinja/issues/392). For CKAN we have to take care that the template helper functions deal with this correctly: Either they need to support both `str` and `unicode` input or we need to make sure to convert string literals in templates to `unicode` (perhaps using a custom `u` filter, i.e. `"föö"|u`). This also holds for any other functions that are called by the templates (e.g. `_`) and for the CKAN Jinja2 extensions (`snippet`, etc.).

#### PyBabel
* The [official documentation](http://babel.pocoo.org/en/latest/dev.html#unicode) is rather vague regarding Unicode handling:

  > Internally everything is unicode that makes sense to have as unicode. The exception to this rule are things which on Python 2 traditionally have been bytes. For example file names on Python 2 should be treated as bytes wherever possible.
  >
  > Encode / decode at boundaries explicitly. Never assume an encoding in a way it cannot be overridden. utf-8 should be generally considered the default encoding.

  In my interpretation, this means that we should get `unicode` translations if we pass `unicode` inputs.
*  The docs also say not to use `unicode_literals`, but it seems that this is a rule for the development of PyBabel itself (no specific reason is given why using `unicode_literals` in code that *uses* PyBabel is a bad idea).
* I couldn't find any information regarding the encoding of message catalogs, but as usual UTF8 seems like the best choice. We need to check whether this is automatically done by the extraction process.

#### Pylons
* The [Pylons documentation on Unicode](http://docs.pylonsproject.org/projects/pylons-webframework/en/latest/tutorials/understanding_unicode.html) says

  > Pylons automatically coerces incoming form parameters (request.POST, GET (quote GET) and params) into unicode objects (as of Pylons 0.9.6).

  However:

  > Only parameter values (not their associated names) are decoded to unicode by default. Since parameter names commonly map directly to Python variable names (which are restricted to the ASCII character set), it’s usually preferable to handle them as strings. For example, passing form parameters to a function as keyword arguments (e.g. **request.params.mixed()) doesn’t work with unicode keys.
  >
  > To make WSGIRequest decode parameter names anyway, enable the decode_param_names option on either the WSGIRequest object or the request_settings dictionary. FieldStorage's name attributes are also decoded to unicode when this option is enabled.

  Since the issue of using `unicode` keywords for parameters has been [fixed long ago](http://stackoverflow.com/a/4598644/857390) it seems that enabling `decode_param_names` would be a good idea.
* Regarding file uploads, the docs say

  > FieldStorage (file upload) objects will be handled specially for unicode parameters: what’s provided is a copy of the original FieldStorage object with a unicode version of its filename attribute.

  This is fine (although we should pass filenames to IO-functions as `str` we should still store them as `unicode` internally). I couldn't find information regarding the type of uploaded *content* (i.e. the upload's [`file` and `value` properties](http://docs.pylonsproject.org/projects/pylons-webframework/en/latest/forms.html#file-uploads)). That needs to be checked.
* Regarding output encoding, the docs say that you should add the necessary HTML `<meta>` tag and `Content-Type` HTTP header.

#### SQLAlchemy
* SQLAlchemy provides the [`Unicode`](http://docs.sqlalchemy.org/en/latest/core/type_basics.html#sqlalchemy.types.Unicode) and [`UnicodeText`](http://docs.sqlalchemy.org/en/latest/core/type_basics.html#sqlalchemy.types.UnicodeText) column types that we already use. These accept and return `unicode`. If they receive `str` they issue a warning and convert to `unicode`.
* The default PostgreSQL-driver for SQLAlchemy is [psycopg2](http://initd.org/psycopg/). [SQLAlchemy's psycopg2 documentation](http://docs.sqlalchemy.org/en/latest/dialects/postgresql.html#unicode-with-psycopg2) says that it uses the databases encoding by default (which is UTF8 for CKAN). Regardless of the encoding, strings are exchanged as Unicode with the DBAPI.

**Comment by wardi** [comment]
@torfsen are you waiting on further guidance here? It sounds like you have a good handle on the trade-offs.

**Comment by torfsen** [comment]
@wardi: I think I'm fine regarding the technical issues and will start working on them once I find the time.

However, I'm not sure about the best way to contribute the changes, since this will touch a lot of code all over the place and likely is too much work to do in a single short burst. So I could either do multiple PRs (one after another) or I could maintain a WIP PR for a longer time and ask for opinions on the way while rebasing from master regularly to keep things in sync. I think I'd prefer the latter approach, but I'm open to suggestions.

**Comment by wardi** [comment]
I'd prefer this broken into separate PRs so we can get the necessary changes in sooner and discuss  specific approaches separately.
